src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4464,64 +4464,99 @@ void C2_MacroAssembler::vector_maskall_operation(KRegister dst, Register src, in
-    }
-    vpand(xtmp3, src, xtmp1, vec_enc);
-    vpshufb(xtmp3, xtmp2, xtmp3, vec_enc);
+    vector_popcount_byte(xtmp1, src, dst, xtmp2, rtmp, vec_enc);
+    vbroadcastd(xtmp2, rtmp, 0x00FF00FF, vec_enc);
+    vpsrlw(dst, xtmp1, 8, vec_enc);
+    vpand(xtmp1, xtmp1, xtmp2, vec_enc);
+    vpaddw(dst, dst, xtmp1, vec_enc);
+  }
+}
+
+void C2_MacroAssembler::vector_popcount_byte(XMMRegister dst, XMMRegister src, XMMRegister xtmp1,
+                                             XMMRegister xtmp2, Register rtmp, int vec_enc) {
+  if (VM_Version::supports_avx512_bitalg()) {
+    vpopcntb(dst, src, vec_enc);
+  } else {
+    assert((vec_enc == Assembler::AVX_512bit && VM_Version::supports_avx512bw()) || VM_Version::supports_avx2(), "");
+    vbroadcastd(xtmp1, rtmp, 0x0F0F0F0F, vec_enc);
     vpsrlw(dst, src, 4, vec_enc);
     vpand(dst, dst, xtmp1, vec_enc);
+    vpand(xtmp1, src, xtmp1, vec_enc);
+    vmovdqu(xtmp2, ExternalAddress(StubRoutines::x86::vector_popcount_lut()), rtmp, vec_enc);
+    vpshufb(xtmp1, xtmp2, xtmp1, vec_enc);
     vpshufb(dst, xtmp2, dst, vec_enc);
-    vpaddb(xtmp3, dst, xtmp3, vec_enc);
-    vpxor(xtmp1, xtmp1, xtmp1, vec_enc);
-    vpunpckhdq(dst, xtmp3, xtmp1, vec_enc);
-    vpsadbw(dst, dst, xtmp1, vec_enc);
-    vpunpckldq(xtmp2, xtmp3, xtmp1, vec_enc);
-    vpsadbw(xtmp2, xtmp2, xtmp1, vec_enc);
-    vpackuswb(dst, xtmp2, dst, vec_enc);
+    vpaddb(dst, dst, xtmp1, vec_enc);
   }
 }
 
-void C2_MacroAssembler::vector_popcount_long(BasicType bt, XMMRegister dst, XMMRegister src, XMMRegister xtmp1,
-                                             XMMRegister xtmp2, XMMRegister xtmp3, Register rtmp, int vec_enc) {
+void C2_MacroAssembler::vector_popcount_long(BasicType rbt, XMMRegister dst, XMMRegister src, XMMRegister xtmp1,
+                                             XMMRegister xtmp2, Register rtmp, int vec_enc) {
   if (VM_Version::supports_avx512_vpopcntdq()) {
     vpopcntq(dst, src, vec_enc);
   } else if (vec_enc == Assembler::AVX_512bit) {
-    assert(VM_Version::supports_avx512bw(), "");
-    movl(rtmp, 0x0F0F0F0F);
-    movdl(xtmp1, rtmp);
-    vpbroadcastd(xtmp1, xtmp1, vec_enc);
-    evmovdqul(xtmp2, k0, ExternalAddress(StubRoutines::x86::vector_popcount_lut()), true, vec_enc, rtmp);
-    vpandq(xtmp3, src, xtmp1, vec_enc);
-    vpshufb(xtmp3, xtmp2, xtmp3, vec_enc);
-    vpsrlw(dst, src, 4, vec_enc);
-    vpandq(dst, dst, xtmp1, vec_enc);
-    vpshufb(dst, xtmp2, dst, vec_enc);
-    vpaddb(xtmp3, dst, xtmp3, vec_enc);
-    vpxorq(xtmp1, xtmp1, xtmp1, vec_enc);
-    vpsadbw(dst, xtmp3, xtmp1, vec_enc);
+    vector_popcount_byte(xtmp1, src, dst, xtmp2, rtmp, vec_enc);
+    vpxorq(xtmp2, xtmp2, xtmp2, vec_enc);
+    vpsadbw(dst, xtmp1, xtmp2, vec_enc);
   } else {
     // We do not see any performance benefit of running
     // above instruction sequence on 256 bit vector which
     // can operate over maximum 4 long elements.
     ShouldNotReachHere();
   }
-  if (bt == T_INT) {
+  if (rbt == T_INT) {
```

Comment:
So the SLP and the VectorAPI uses the same `PopCountVL` with different basic type?

Reply:
Is it possible to separate this cast for SLP and always uses "T_LONG' as the basic type for `PopCountVL` ?

Reply:
We can have type homogeneity for PopCountVL across VectorAPI and SLP once we start auto-vectorizing ConvL2I  operation.

Reply:
I think that makes the code hard to maintain, and I think @merykitty 's unaddressed comment in: https://github.com/openjdk/jdk/pull/6857#discussion_r780917717 is a preferable approach. 

FYI: @fg1417 has a ConvL2I auto-vectorization support under review: https://github.com/openjdk/jdk/pull/7806.

Reply:
Yeah, so if the `ConvL2I` is supported to be auto-vectorized after then, the `PopCountL` is better to be auto-vectorized to "PopCountVL + VectorCastL2X" ? Do you have any plan on it? Thanks so much！

Reply:
> Yeah, so if the `ConvL2I` is supported to be auto-vectorized after then, the `PopCountL` is better to be auto-vectorized to "PopCountVL + VectorCastL2X" ? Do you have any plan on it? Thanks so much！

Agree, with your proposal.

Reply:
OK, thanks!

Reply:
> > Yeah, so if the `ConvL2I` is supported to be auto-vectorized after then, the `PopCountL` is better to be auto-vectorized to "PopCountVL + VectorCastL2X" ? Do you have any plan on it? Thanks so much！
> 
> Agree to your proposal, once we have auto-vectorization for ConvL2I n place we can change the IR.

Here is an interim patch which type casts the result of PopCountVL into integer vector and thus prevent any special 
handling in target specific backend implementations. But it seems more appropriate to generate correct IR snippet (PopCountL + ConvL2I) during scalar operation inline expansion and explicitly handle ConvL2I during auto-vectorization.  

```
diff --git a/src/hotspot/share/opto/superword.cpp b/src/hotspot/share/opto/superword.cpp
index 1bf4c7a3282..b3dd90d1867 100644
--- a/src/hotspot/share/opto/superword.cpp
+++ b/src/hotspot/share/opto/superword.cpp
@@ -2549,11 +2549,21 @@ void SuperWord::output() {
           vn = VectorNode::make(opc, in1, in2, vlen, velt_basic_type(n));
           vlen_in_bytes = vn->as_Vector()->length_in_bytes();
         }
+      } else if (opc == Op_PopCountL) {
+        // TODO: Succeed PopCountL by ConvL2I during initial graph construction 
+        // once auto-vectorizer supports ConvL2I operation.
+        assert(n->req() == 2, "only one input expected");
+        Node* in = vector_opd(p, 1);
+  	   BasicType bt = in->bottom_type()->is_vect()->element_basic_type();
+         in = VectorNode::make(opc, in, NULL, vlen, bt);
+         _igvn.register_new_node_with_optimizer(in);
+	   vn = VectorCastNode::make(Op_VectorCastL2X, in, T_INT, vlen);
+         vlen_in_bytes = vn->as_Vector()->length_in_bytes();
       } else if (opc == Op_SqrtF || opc == Op_SqrtD ||
                  opc == Op_AbsF || opc == Op_AbsD ||
                  opc == Op_AbsI || opc == Op_AbsL ||
                  opc == Op_NegF || opc == Op_NegD ||
-                 opc == Op_PopCountI || opc == Op_PopCountL) {
+                opc == Op_PopCountI) {
         assert(n->req() == 2, "only one input expected");
         Node* in = vector_opd(p, 1);
         vn = VectorNode::make(opc, in, NULL, vlen, velt_basic_type(n));
```

Reply:
Yeah, thanks for it. I will do it once PR https://github.com/openjdk/jdk/pull/7806 is merged. Thanks!